File: main/finish/workflow_candidates/gersteinlab__ASTRO/bash/GETmat/runBowtie.pl
use strict;
use warnings;
use Getopt::Long;


my ($referencefile, $inputfa, $outputfile, $thread);
GetOptions(
    "r|referencefile=s" => \$referencefile,
    "i|inputfa=s"       => \$inputfa,
    "o|outputfile=s"    => \$outputfile,
    "t|thread=s"        => \$thread
);
$thread //= '8'; 


my @cmd = (
    'bowtie2', '-p', $thread, '-x', $referencefile, '-U', $inputfa, '-D', '20', '-R', '3', '-N', '1', '-L', '6',
    '-i', 'C,5', '--score-min', 'C,10', '--no-hd', '--no-sq', '-a', '--ma', '1', '--rdg', '0,1', '--rfg', '0,1',
    '--mp', '1,1', '--gbar', '1', '--local'
);


open my $process, '-|', @cmd or die "Could not start process: $!";


open my $output_file, '>', $outputfile or die "Could not open file '$outputfile': $!";
my ($major_read, $major_map, $major_value, $see1);

$see1 = 0;
while (my $line = <$process>) {

    chomp $line;
    my @parts = split(/\s+/, $line);
    my $current_read = $parts[0];
    my $current_map = $parts[2];
    my $current_value;
    if ($current_map ne '*') {
        unless ( ($current_value) = $parts[11] =~ /^AS:i:(\d+)$/) {
            print STDERR "Error in line: $line\n";
            die "Tag string does not match the required format AS:i:x";
        }
    }else{
        print $output_file "$current_read\t*\n";
        next;
    }

    if (!defined $major_read || $current_read ne $major_read) {
        print $output_file "$major_read\t$major_map\n" if defined $major_read;
        $see1 = 1;
        $major_read = $current_read;
        $major_value = $current_value;
        $major_map = $parts[2];
    }else{
        if ($see1) {
            if ($major_value <= $current_value) {
                $major_map = '*';
            }
            $see1 = 0;
        }
    }
}

if (defined $major_read) {
    print $output_file "$major_read\t$major_map\n";
}
close $output_file;
close $process;

File: main/finish/workflow_candidates/gersteinlab__ASTRO/python/scripts/runBowtie.py
import subprocess
import re
import argparse


ap = argparse.ArgumentParser()
ap.add_argument("-r", "--referencefile", required=True, help="Path to input BAM file")
ap.add_argument("-i", "--inputfa", required=True, help="Path to input BAM file")
ap.add_argument("-o", "--outputfile", required=True, help="Path to input BAM file")
ap.add_argument(
    "-t", "--thread", required=False, default="8", type=str, help="Number of threads"
)
args = vars(ap.parse_args())

outputfile = args["outputfile"]
inputfa = args["inputfa"]
referencefile = args["referencefile"]
thread = args["thread"]


cmd = [
    "bowtie2",
    "-p",
    thread,
    "-x",
    referencefile,
    "-U",
    inputfa,
    "-D",
    "7",
    "-R",
    "2",
    "-N",
    "1",
    "-L",
    "6",
    "-i",
    "C,7",
    "--score-min",
    "C,10",
    "--no-hd",
    "--no-sq",
    "-a",
    "--ma",
    "1",
    "--rdg",
    "0,1",
    "--rfg",
    "0,1",
    "--mp",
    "1,1",
    "--gbar",
    "1",
    "--local",
]


process = subprocess.Popen(cmd, stdout=subprocess.PIPE, text=True)

with open(outputfile, "w") as output_file:
    major_read = None
    major_map = None
    major_value = None
    see1 = 0
    for line in process.stdout:
        parts = line.strip().split()
        current_read = parts[0]
        match = re.match(r"^AS:i:(\d+)$", parts[11])
        current_map = parts[2]
        if current_map != "*":
            if match:
                current_value = int(match.group(1))
            else:
                print("Error in line:", line)
                raise ValueError("Tag string does not match the required format AS:i:x")
        else:
            output_file.write(f"{major_read}\t*" + "\n")
            continue
        if current_read != major_read:
            if major_read is not None:
                output_file.write(f"{major_read}\t{major_map}" + "\n")
            see1 = 1
            major_read = current_read
            major_value = current_value
            major_map = current_map
        else:
            if see1 == 1:
                if major_value <= current_value:
                    major_map = "*"
                see1 = 0
    if major_read is not None:
        output_file.write(f"{major_read}\t{major_map}" + "\n")


process.wait()
